### PR TITLE
gh-116647: Fix recursive child in dataclasses

### PR DESCRIPTION
**Pull request by ghost**
https://github.com/python/cpython/issues/116647

* Issue: gh-116647

### CODE REVIEW

File: Lib/dataclasses.py [hunk 1]
```diff
@@ -1073,7 +1073,7 @@ def _process_class(cls, init, repr, eq, order, unsafe_hash, frozen,
         # Create __eq__ method.  There's no need for a __ne__ method,
         # since python will call __eq__ and negate it.
         cmp_fields = (field for field in field_list if field.compare)
-        terms = [f'self.{field.name}==other.{field.name}' for field in cmp_fields]
+        terms = [f'(self.{field.name},)==(other.{field.name},)' for field in cmp_fields]
```
Comment (ericvsmith) [inline Lib/dataclasses.py]: Is this an improvement over just reverting to the code before 18cfc1eea569f0ce72ad403840c0e6cc5f81e1c2?

Comment (ghost) [reply Lib/dataclasses.py]: I think this is just reverting to the code before https://github.com/python/cpython/commit/18cfc1eea569f0ce72ad403840c0e6cc5f81e1c2 

due to `(object_a, ) == (object_b)` will first compare ids of objects, if it's not same then call __eq__ of the object, that may not cause `RecursionError: maximum recursion depth exceeded` when the objects pointers are the same

but object_a == object_b directly calls object.__eq__ whether the pointers is same or not, so i think remove https://github.com/python/cpython/commit/18cfc1eea569f0ce72ad403840c0e6cc5f81e1c2 will solve the errors and provides better performance when the child have the same pointers will cause early return without calling __eq__ of objects

Comment (ericvsmith) [reply Lib/dataclasses.py]: My concern is that proposed code builds lots of tuples. The idea of https://github.com/python/cpython/commit/18cfc1eea569f0ce72ad403840c0e6cc5f81e1c2 was to improve efficiency, and if the proposed code isn't at least as efficient as the pre-https://github.com/python/cpython/commit/18cfc1eea569f0ce72ad403840c0e6cc5f81e1c2 code, then it's a step backward.

File: Lib/test/test_dataclasses/__init__.py [hunk 2]
```diff
@@ -2471,6 +2471,15 @@ def __repr__(self):
 
 
 class TestEq(unittest.TestCase):
+    def test_recursive_eq(self):
```
Comment (Jason-Y-Z) [inline Lib/test/test_dataclasses/__init__.py:2474]: Thanks for the change! Per discussions in the issue, maybe also worth adding a test case for mutual recursive classes (given that's intended to be included in the scope of this PR)?

**Review by Jason-Y-Z — COMMENTED** [review]
Overall LGTM, a small comment on the test case

File: Lib/test/test_dataclasses/__init__.py (hunk 2 above)
Comment (ghost) [reply Lib/test/test_dataclasses/__init__.py:2474]: Thanks for the comment, but I don't think mutual recursive classes can be implement with good performance, I've think ways like maintenance dict like id(obj) as key to check if it's recursive but it'll cost spaces and times to make compare like that.

File: Lib/dataclasses.py (hunk 1 above)
Comment (ghost) [reply Lib/dataclasses.py]: Thanks for the comment, I've improved the efficiency base on https://github.com/python/cpython/commit/18cfc1eea569f0ce72ad403840c0e6cc5f81e1c2 now it'll compare id(obj) first to make sure if the child point to themselves will not cause maximum recursion depth and it'll be faster when compare exact same objects.

File: Lib/dataclasses.py [hunk 3]
```diff
@@ -1075,7 +1075,9 @@ def _process_class(cls, init, repr, eq, order, unsafe_hash, frozen,
         cmp_fields = (field for field in field_list if field.compare)
         terms = [f'self.{field.name}==other.{field.name}' for field in cmp_fields]
         field_comparisons = ' and '.join(terms) or 'True'
-        body =  [f'if other.__class__ is self.__class__:',
+        body =  [f'if id(self) == id(other):',
```
Comment (encukou) [inline Lib/dataclasses.py]: To avoid two function calls, use the `is` operator:
```suggestion
        body =  [f'if self is other:',
```

Comment (ghost) [reply Lib/dataclasses.py]: Thanks for the comment, I've changed the code style.

**Review by carljm — APPROVED** [review]

File: Lib/test/test_dataclasses/__init__.py (hunk 2 above)
Comment (Jason-Y-Z) [reply Lib/test/test_dataclasses/__init__.py:2474]: Make sense, thanks for the clarification!

**Pull request merged**